halo2_proofs/src/plonk/mv_lookup/prover.rs

```diff
@@ -0,0 +1,440 @@
+    compressed_table_expression: Polynomial<C::Scalar, LagrangeCoeff>,
+    m_values: Polynomial<C::Scalar, LagrangeCoeff>,
+}
+
+#[derive(Debug)]
+pub(in crate::plonk) struct Committed<C: CurveAffine> {
+    pub(in crate::plonk) m_poly: Polynomial<C::Scalar, Coeff>,
+    pub(in crate::plonk) phi_poly: Polynomial<C::Scalar, Coeff>,
+}
+
+pub(in crate::plonk) struct Evaluated<C: CurveAffine> {
+    constructed: Committed<C>,
+}
+
+impl<F: FieldExt> Argument<F> {
+    pub(in crate::plonk) fn prepare<
+        'a,
+        'params: 'a,
+        C,
+        P: Params<'params, C>,
+        E: EncodedChallenge<C>,
+        R: RngCore,
+        T: TranscriptWrite<C, E>,
+    >(
+        &self,
+        pk: &ProvingKey<C>,
+        params: &P,
+        domain: &EvaluationDomain<C::Scalar>,
+        theta: ChallengeTheta<C>,
+        advice_values: &'a [Polynomial<C::Scalar, LagrangeCoeff>],
+        fixed_values: &'a [Polynomial<C::Scalar, LagrangeCoeff>],
+        instance_values: &'a [Polynomial<C::Scalar, LagrangeCoeff>],
+        challenges: &'a [C::Scalar],
+        mut rng: R, // in case we want to blind (do we actually need zk?)
+        transcript: &mut T,
+    ) -> Result<Prepared<C>, Error>
+    where
+        C: CurveAffine<ScalarExt = F>,
+        C::Curve: Mul<F, Output = C::Curve> + MulAssign<F>,
+    {
+        // Closure to get values of expressions and compress them
+        let compress_expressions = |expressions: &[Expression<C::Scalar>]| {
+            let compressed_expression = expressions
+                .iter()
+                .map(|expression| {
+                    pk.vk.domain.lagrange_from_vec(evaluate(
+                        expression,
+                        params.n() as usize,
+                        1,
+                        fixed_values,
+                        advice_values,
+                        instance_values,
+                        challenges,
+                    ))
+                })
+                .fold(domain.empty_lagrange(), |acc, expression| {
+                    acc * *theta + &expression
+                });
+            compressed_expression
+        };
+
+        // Get values of input expressions involved in the lookup and compress them
+        let compressed_inputs_expressions: Vec<_> = self
+            .inputs_expressions
+            .iter()
+            .map(|input_expressions| compress_expressions(input_expressions))
+            .collect();
+
+        // Get values of table expressions involved in the lookup and compress them
+        let compressed_table_expression = compress_expressions(&self.table_expressions);
+
+        let blinding_factors = pk.vk.cs.blinding_factors();
+
+        // compute m(X)
+        let table_index_value_mapping: BTreeMap<C::Scalar, usize> = compressed_table_expression
+            .iter()
+            .take(params.n() as usize - blinding_factors - 1)
+            .enumerate()
+            .map(|(i, &x)| (x, i))
+            .collect();
```

Comment:
We need to run more benchmark on this op as constructing BTreeMap over a large table might take a lot of time.

Reply:
This is fixed in #71 (commit https://github.com/scroll-tech/halo2/commit/d901014b088601b994a427a22ef280f217214b5a).